### PR TITLE
[TILES-PW-2] PLU-672: crypto-related helpers

### PR DESCRIPTION
**Pull request by pregnantboy**
## Changes

Added authentication utilities for tile password hashing and view token management with JWT-based session tokens.

### What changed?

Created `auth-tiles.ts` helper module with functions for:
- Password hashing using scrypt with tiles viewOnlyKey as salt (`hashTilePassword`, `verifyTilePassword`)
- Random token nonce generation (`generateTokenNonce`)
- JWT view token creation and verification (`generateViewToken`, `verifyViewToken`)

The view tokens include tileId, viewOnlyKey, and tokenNonce in the payload and expire after 24 hours. Password hashes are automatically invalidated when the viewOnlyKey or passwordNonce changes since it's used as the salt.

### How to test?

Covered by unit tests
- Correct and incorrect password verification
- Valid and invalid view token scenarios (wrong parameters, malformed tokens, expired tokens, wrong signing secret)
- Edge cases like non-JWT errors

### CODE REVIEW

**Comment by pregnantboy** [comment]
* **#1463** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1463?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1459** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1459?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1458** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1458?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1457** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1457?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1456** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1456?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1452** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1452?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1451** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1451?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1450** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1450?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1450?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1449** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1449?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `develop-v2`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/backend/src/helpers/auth-tiles.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+export function verifyTilePassword(
+  password: string,
+  storedHash: string,
+  viewOnlyKey: string,
+): boolean {
+  const inputHash = scryptSync(password, viewOnlyKey, SCRYPT_KEY_LENGTH)
+  const storedHashBuffer = Buffer.from(storedHash, 'base64')
+  return timingSafeEqual(inputHash, storedHashBuffer)
+}
```
[the remark this answers is not in the document]
Comment (kevinkim-ogp) [reply packages/backend/src/helpers/auth-tiles.ts:35]: ~~makes sense~~
actually very unlikely to happen, but could still add for peace of mind

**Review by kevinkim-ogp — APPROVED** [review]
other than the comment from graphite, lgtm!

**Comment by pregnantboy** [comment]
### Merge activity

* **Mar 10, 6:25 AM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/opengovsg/plumber/1450).
* **Mar 10, 6:27 AM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/opengovsg/plumber/1450).
* **Mar 10, 6:29 AM UTC**: [Graphite](https://app.graphite.com/github/pr/opengovsg/plumber/1450) rebased this pull request as part of a merge.
* **Mar 10, 6:29 AM UTC**: The [Graphite](https://app.graphite.com/github/pr/opengovsg/plumber/1450) merge of this pull request was cancelled.
* **Mar 10, 6:29 AM UTC**: @pregnantboy merged this pull request with [Graphite](https://app.graphite.com/github/pr/opengovsg/plumber/1450).
